keps/sig-node/4381-dra-numeric-parameters/README.md

```diff
@@ -0,0 +1,1054 @@
+	Kind string
+
+	// Shareable is copied into the AllocationResult when a claim
+	// gets allocated.
+	Shareable bool
+}
+```
+
+### NodeResources
+
+For each node, one or more NodeResources objects gets created:
+
+```
+// NodeResources provides information about available
+// resources on individual nodes.
+type NodeResources struct {
+	metav1.TypeMeta
+	// Standard object metadata
+	// +optional
+	metav1.ObjectMeta
+
+	// NodeName identifies the node where the capacity is available.
+	// A field selector can be used to list only NodeResources
+	// objects with a certain node name.
+	NodeName string
+
+	// DriverName identifies the DRA driver providing the capacity information.
+	// A field selector can be used to list only NodeResources
+	// objects with a certain driver name.
+	DriverName string
+
+	// Instances describes discrete resources managed by the driver on
+	// the node. It is possible to create more than one NodeResources
+	// object for the same node and driver if this array becomes to large
+	// for a single object.
+	Instances []NodeResourceInstance
+}
+
+// NodeResourceInstance describes one discrete resource instance.
+type NodeResourceInstance struct {
+	// ID is chosen by the driver to distinguish between different resource
+	// instances. It only needs to be unique for the driver and node.  In
+	// other words, the tuple of NodeName, DriverName, InstanceID needs to
+	// be unique in the cluster.
+	ID string
+
+    // Attributes describe the instance.
+	Attributes []InstanceAttributes
+}
+
+// InstanceAttributes must have one and only one field set.
+type InstanceAttributes struct {
+	<one pointer per semantic model>
+}
+```
+
+If a driver needs to reduce resource capacity, then there is a risk that a claim gets
+allocated using that capacity while the kubelet is updating the
+NodeResources object. The implementations of semantic models must handle
+scenarios where more resources are allocated than available. The DRA driver
+kubelet plugin must double-check that the allocated resources are still
+available when NodePrepareResource is called. If not, the pod cannot
+start until the resource comes back.
+
+### Builtin controllers
+
+`k8s.io/dynamic-resource-allocation/builtincontroller` defines Go interfaces
+that need to be implemented by a semantic model to handle tracking of resource
+usage and allocation/deallocation. The package also has a global registry of
+builtin controllers.
+
+##### Initialization
+
+All registered builtin controllers get activated when kube-scheduler or CA
+start. At this point they start building an internal model of which resources
+are available (based on those entries in NodeResources objects that they
+are responsible for) and how those resources are already used (based on the
+status of allocated claims).
+
+##### Allocation
+
+When the dynamic resource plugin in kube-scheduler encounters a claim which,
+according to the class, uses semantic parameters, it expands the CEL template
```

Comment:
ok, yeah, this looks like it hasn't yet been updated for the latest discussions (no CEL, use a driver controller for the translation).

From the scheduler (plugin) perspective, it just needs to find the `ClaimParameter` resources that correspond to the pod it is trying to schedule, then use the selection criteria to reduce the set of possible nodes, then use that smaller set to do the fit calculation against the capacity availability. I think we need a description here as to how this works (or references back to the main DRA KEP).

Reply:
I think the lines below have yet to be updated, so I am going to stop for tonight.

Reply:
Updated now.